### ISSUE TITLE
Trading Dashboard — Multi-Page Web UI with Docker ARM64 Deployment

### ISSUE THREAD

**Issue by Pr0mp7** [Photonensauger/Quant-2.0#4]
## Summary

Build a professional multi-page trading dashboard for Quant 2.0 that provides real-time portfolio monitoring, backtest visualization, model analytics, and risk management views. The dashboard runs as a containerized Python application with **ARM64 support for Raspberry Pi deployment**.

---

## Motivation

Currently, all interaction with Quant 2.0 happens through CLI scripts (`run_backtest.py`, `train_model.py`, `live_trade.py`). There is no visual way to:
- Monitor portfolio performance and live positions
- Compare backtest results across strategies and parameters
- Analyze individual model predictions and confidence levels
- Track risk exposure and drawdown in real-time
- View system health (data freshness, execution latency)

A web-based dashboard solves all of these and makes the system accessible to non-CLI users.

---

## Tech Stack

- **Framework:** Dash 4.0 (Plotly) + dash-bootstrap-components
- **Design:** Once UI dark theme (custom CSS + Dash 4.x `--Dash-*` variable overrides)
- **Charts:** Plotly 6.x
- **Fonts:** Plus Jakarta Sans (UI) + JetBrains Mono (data)
- **Server:** Gunicorn (2 workers)
- **Deployment:** Docker (multi-stage ARM64 build), K3s compatible

---

### Page 1: Portfolio Overview (Home) ✅
KPI cards (return, Sharpe, max DD, win rate, portfolio value, daily P&L), equity curve, drawdown chart, monthly returns heatmap, return distribution histogram. Backtest selector dropdown.

### Page 2: Positions & Trades ✅
Active positions section, trade log table with P&L, P&L by trade bar chart, cumulative P&L line chart. Backtest selector dropdown. **CSV/JSON export buttons for trade log.**

### Page 3: Backtest Explorer ✅
Backtest selector with KPI row (return, Sharpe, max DD, win rate, trades, profit factor). Equity curve with trade entry/exit markers (triangles). Full 16-metric table. Trade summary (wins/losses, best/worst). Multi-backtest comparison with normalized equity overlay. **CSV/JSON export buttons for metrics and trade log. Interactive Run Backtest form (strategy, assets, interval, dates, capital).**

### Page 4: Model Analytics ✅
Summary KPIs (total models, trained count, backtests). Model registry cards with checkpoint status (trained/not, size, date). Backtest performance bar chart. Side-by-side metrics comparison table (best values highlighted). **Interactive Train Models form (model selector, assets, interval, days, epochs, batch size) with subprocess execution, live log output, and status indicator.**

### Page 5: Risk Management ✅
8 KPI cards (Sharpe, Sortino, Calmar, VaR 95%, CVaR 95%, volatility, max DD, DD duration). Rolling metrics chart (21-day vol + Sharpe, dual Y-axis). Drawdown periods visualization. VaR distribution with VaR/CVaR lines. Backtest selector dropdown. **Configurable risk threshold alerts with colored banner (amber/red).**

### Page 6: System Health ✅
Data freshness, model checkpoints, system info.

---

### Project Structure
```
dashboard/
├── app.py                  # Dash app entry point
├── config.py               # Settings, ASSET_OPTIONS, RISK_THRESHOLDS
├── assets/
│   └── style.css           # Once UI dark theme + responsive breakpoints
├── pages/
│   ├── overview.py         # Page 1: Portfolio Overview
│   ├── positions.py        # Page 2: Positions & Trades
│   ├── backtest.py         # Page 3: Backtest Explorer
│   ├── models.py           # Page 4: Model Analytics
│   ├── risk.py             # Page 5: Risk Management
│   └── system.py           # Page 6: System Health
├── components/
│   ├── kpi_card.py         # Reusable metric card component
│   ├── equity_chart.py     # Equity curve + benchmark chart
│   ├── drawdown_chart.py   # Underwater drawdown chart
│   ├── trade_table.py      # Dash DataTable with sort/filter/pagination
│   ├── heatmap.py          # Monthly returns heatmap
│   └── navbar.py           # Top header pill navigation
├── data/
│   ├── cache.py            # TTLCache with max_size eviction + stats()
│   └── loader.py           # Parquet + JSON data loader with dual TTL caches
├── k8s/                    # Kubernetes manifests
├── Dockerfile              # Multi-stage ARM64 build
├── requirements.txt        # Dashboard-specific deps
└── README.md               # Setup, deploy, architecture docs
```

---

### Phase 1: Foundation ✅
- [x] Set up `dashboard/` directory structure
- [x] Initialize Dash app with multi-page routing
- [x] Create navbar component (top header pill navigation)
- [x] Set up Once UI dark theme with custom CSS
- [x] Create reusable KPI card component
- [x] Implement `data/loader.py` to read Parquet files and backtest results
- [x] Write Dockerfile (multi-stage, ARM64-compatible)
- [x] Add `dashboard` optional dependency group to `pyproject.toml`
- [x] Deploy and verify on K3s

### Phase 2: Core Pages ✅
- [x] **Page 1: Portfolio Overview** — KPI cards, equity curve, drawdown, monthly heatmap, return distribution
- [x] **Page 2: Positions & Trades** — trade table, P&L bar chart, cumulative P&L
- [x] **Page 3: Backtest Explorer** — selector, full metrics, equity with trade markers, multi-backtest comparison

### Phase 3: Advanced Pages ✅
- [x] **Page 4: Model Analytics** — registry with checkpoint status, backtest comparison chart + table
- [x] **Page 5: Risk Management** — VaR/CVaR, rolling vol/Sharpe, drawdown periods, VaR distribution
- [x] **Page 6: System Health** — data freshness, model checkpoints, system info

### Phase 4: Interactive Features ✅
- [x] CSV/JSON export buttons for trade log (Backtest Explorer + Positions pages)
- [x] JSON export for full backtest metrics (Backtest Explorer page)
- [x] Run Backtest from GUI — interactive form with strategy, assets, interval, dates, capital inputs
- [x] Subprocess execution with live stdout log output and status indicator
- [x] Auto-reload backtest list on completion
- [x] Docker image updated with torch CPU, yfinance, scikit-learn, python-dotenv
- [x] K8s deployment memory limit increased to 1Gi (torch requirement)
- [x] **Train Models from GUI** — model selector (all/transformer/itransformer/lstm/momentum), assets multi-select, interval, training days, epochs, batch size
- [x] **Train subprocess management** — file-based IPC, status badge (idle/running/complete/failed), live log output
- [x] **Shared ASSET_OPTIONS** — moved to `config.py` for DRY reuse across backtest and models pages
- [x] **Backtest form UI fixes** — date inputs min-height 40px, Assets dropdown flex:2, dropdown z-index/max-height

### Phase 5: Polish & Production ✅
- [x] Trade log sortable and filterable (Dash DataTable with native sort/filter/pagination)
- [x] Configurable risk threshold alerts (amber/red banner, KPI card color changes)
- [x] Performance optimization (TTLCache max_size eviction, 300s backtest TTL, stats/len methods)
- [x] Mobile-responsive layout (1024/768/480px breakpoints, table overflow, form collapse)
- [x] Integration tests (15/15 passing — TTLCache + DashboardDataLoader)
- [x] Documentation (dashboard/README.md — setup, deploy, architecture)
- [ ] CI/CD: GitHub Actions workflow for multi-arch Docker build (deferred — manual build flow)

---

## Acceptance Criteria

- [x] Dashboard launches and serves on ARM64
- [x] All 6 pages render correctly with sample data
- [x] Equity curve displays with trade entry/exit markers
- [x] Trade log is filterable and sortable
- [x] Backtest results display with full metrics and comparison view
- [x] Risk metrics (VaR, CVaR, Sharpe, rolling vol, drawdown) display correctly
- [x] Dark theme is consistent across all pages (Once UI + Dash 4.x token overrides)
- [x] Health check endpoint responds on `/`
- [x] Data refreshes automatically at configured intervals
- [x] Model checkpoint detection and status display
- [x] CSV/JSON export for trade log and metrics
- [x] Run Backtest directly from dashboard GUI
- [x] Train Models directly from dashboard GUI

---

## Dependencies on Existing System

| Dashboard Need | Quant Module | Status |
|----------------|-------------|--------|
| Market data (OHLCV) | `quant.data.storage.DataStorage` (Parquet) | ✅ Ready |
| Backtest metrics | `quant.backtest.metrics.compute_metrics()` | ✅ Ready |
| Backtest execution | `quant.backtest.engine.BacktestEngine` | ✅ Ready |
| Risk management | `quant.portfolio.risk.RiskManager` | ✅ Ready |
| Position sizing | `quant.portfolio.position.PositionSizer` | ✅ Ready |
| Model inference | `quant.models.*` (4 Transformers + PPO) | ✅ Ready |
| Strategy signals | `quant.strategies.MLSignalStrategy` | ✅ Ready |
| Plotly | Already in `pyproject.toml` | ✅ Ready |
| PyTorch CPU | Dashboard Dockerfile (--extra-index-url cpu) | ✅ Ready |
| yfinance | Dashboard requirements.txt | ✅ Ready |

**Comment by Pr0mp7** [comment]
## Phase 5: Polish & Production — Complete ✅

All Phase 5 items implemented and deployed to Pi5 K3s (`https://quant.lan`).

### Changes

**1. Sortable/Filterable Trade Log**
- Replaced static HTML table with `dash_table.DataTable`
- Native column sorting (`sort_action="native"`) and per-column text filters (`filter_action="native"`)
- Pagination at 25 rows, P&L color coding (green ≥0, red <0), Side color coding (LONG/SHORT)
- Filter inputs styled to match Once UI dark theme

**2. Risk Threshold Alerts**
- Added `RISK_THRESHOLDS` config (Sharpe min, drawdown warn/critical, volatility warn, VaR warn)
- Colored alert banner: amber for warnings, red for critical breaches
- KPI card colors now reflect threshold status (e.g. Sharpe < 0.5 → warning amber)

**3. Performance Optimization (Caching)**
- `TTLCache` now supports `max_size` (default 50) with oldest-entry eviction
- Added `__len__()` and `stats()` methods for debugging
- Backtest results use separate cache with 300s TTL (market data stays at 60s)

**4. Mobile-Responsive Layout**
- New 1024px breakpoint for chart height reduction
- 768px: form rows collapse to single column, table overflow-x auto
- 480px: smaller KPI card padding, reduced chart heights
- DataTable filter/pagination dark theme styling

**5. Integration Tests (15/15 passing)**
- `test_cache.py` (8 tests): set/get, TTL expiry, clear, max_size eviction, update existing, len, stats
- `test_loader.py` (7 tests): list_symbols, load_market_data, cache hit, missing file, load_backtest, load_all, clear_cache

**6. Documentation**
- `dashboard/README.md`: pages overview, local dev, env vars, Docker build, Pi5 K3s deploy, architecture tree (82 lines)

### Files Changed (11)
`dashboard/components/trade_table.py`, `dashboard/pages/positions.py`, `dashboard/pages/risk.py`, `dashboard/config.py`, `dashboard/data/cache.py`, `dashboard/data/loader.py`, `dashboard/assets/style.css`, `tests/test_dashboard/conftest.py`, `tests/test_dashboard/test_cache.py`, `tests/test_dashboard/test_loader.py`, `dashboard/README.md`

### Deployment
Built ARM64 on Mac (Colima) → `scp -O` → `k3s ctr images import` → `kubectl rollout restart` → verified 200 OK at `https://quant.lan`

### Remaining from original Phase 5 checklist
- [ ] CI/CD: GitHub Actions workflow for multi-arch Docker build (deferred — no registry, manual build flow works)
